lib/ddtrace/trace_operation.rb

```diff
@@ -0,0 +1,400 @@
+      @service = service
+
+      # State
+      @root_span = nil
+      @active_span = nil
+      @active_span_count = 0
+      @events = events || Events.new
+      @finished = false
+      @spans = []
+    end
+
+    # TODO: Maybe not the right strategy for preventing
+    #       big traces? Prevents new spans from starting if
+    #       too many spans are finished. Has no effect if all
+    #       spans are in-flight... likely needs to change.
+    def full?
+      @max_length > 0 && @active_span_count >= @max_length
+    end
+
+    def finished_span_count
+      @spans.length
+    end
+
+    def finished?
+      @finished == true
+    end
+
+    def sampled?
+      @sampled == true || (!@sampling_priority.nil? && @sampling_priority > 0)
+    end
+
+    def keep!
+      self.sampling_priority = Datadog::Ext::Priority::USER_KEEP
+    end
+
+    def reject!
+      self.sampling_priority = Datadog::Ext::Priority::USER_REJECT
+    end
+
+    def measure(op_name, **span_options, &block)
+      # Don't allow more span measurements if the
+      # trace is already completed. Prevents multiple
+      # root spans with parent_span_id = 0.
+      return yield(SpanOperation.new(op_name), TraceOperation.new) if finished? || full?
+
+      # Create new span
+      span_op = build_span(op_name, **span_options)
+
+      # Start span measurement
+      span_op.measure { |s| yield(s, self) }
+    end
+
+    def build_span(op_name, **span_options)
+      begin
+        # Resolve span options:
+        # Context, parent, service name, etc.
+        span_options = build_span_options(**span_options)
+
+        # Build a new span operation
+        Datadog::SpanOperation.new(op_name, **span_options)
+      rescue StandardError => e
+        Datadog.logger.debug("Failed to build new span: #{e}")
+
+        # Return dummy span
+        SpanOperation.new(op_name)
+      end
+    end
+
+    def flush!
+      finished = finished?
+
+      # Copy out completed spans
+      spans = @spans.dup
+      @spans = []
+
+      # Use them to build a trace
+      build_trace(spans, !finished)
+    end
+
+    # Returns a set of trace headers used for continuing traces.
+    # Used for propagation across execution contexts.
+    # Data should reflect the active state of the trace.
+    def to_digest
+      TraceDigest.new(
+        span_id: (@active_span && @active_span.id) || @parent_span_id,
+        span_name: (@active_span && @active_span.name),
+        span_resource: (@active_span && @active_span.name),
+        span_service: (@active_span && @active_span.service),
+        span_type: (@active_span && @active_span.type),
+        trace_hostname: @hostname,
+        trace_id: @id,
+        trace_name: @name,
+        trace_origin: @origin,
+        trace_process_id: @process_id,
+        trace_resource: @resource,
+        trace_runtime_id: @runtime_id,
+        trace_sampling_priority: @sampling_priority,
+        trace_service: @service
+      ).freeze
```

Comment:
Minor: Not needed, since the `TraceDigest` already `freeze`s itself in the constructor.